Add partition_space to OpenMP

Rombur opened the pull request:
This is the last part of #4935

The PR does the following:
 - add a new constructor for `OpenMP()`
 - use the `ExecutionSpace` in `impl_thread_pool_size()` 
 - use the new `impl_thread_pool_size()` where necessary
 - pass the `ExecutionSpace` by const reference instead of by copy in the default implementation of `partition_space`. This was already done in the backend implementations
 - add `partition_space` for the OpenMP backend. I had to modify a test because the new function emits a warning when the variadic arguments are different types. So you cannot mix `int` and `double` anymore. I don't think that's a big deal and the function is still in `Experimental`.

@dalg24 please do not merge until @crtrott  has time to take a look. It has written the `partition_space` for all the other backends and I would like to have his feedback

dalg24 commented:
I would never merge anything without him looking :)
I requested a review from him

dalg24 commented on core/src/Kokkos_Core.hpp:
```diff
@@ -301,7 +301,7 @@ namespace Experimental {
 //   Default behavior is to return the passed in instance
 template <class ExecSpace, class... Args>
-std::vector<ExecSpace> partition_space(ExecSpace space, Args...) {
+std::vector<ExecSpace> partition_space(ExecSpace const& space, Args...) {
```
This is fine.  I don't think it matters much and the [documentation](https://github.com/kokkos/kokkos-core-wiki/blob/0de2cca848d11ae62f0d1c8c25bc14b17e866285/API-Reference/Kokkos::partition_space.md#interface) was saying it is passed by const reference.

dalg24 commented on core/src/OpenMP/Kokkos_OpenMP_Instance.hpp:
```diff
@@ -139,28 +145,25 @@ inline bool OpenMP::impl_is_initialized() noexcept {
   return Impl::OpenMPInternal::singleton().is_initialized();
 }
 
-inline bool OpenMP::in_parallel(OpenMP const&) noexcept {
-  // FIXME_OPENMP We are forced to use t_openmp_instance because the function is
-  // static and does not use the OpenMP object
-  return ((Impl::OpenMPInternal::singleton().m_level < omp_get_level()) &&
-          (!Impl::t_openmp_instance ||
-           Impl::t_openmp_instance->m_level < omp_get_level()));
+inline bool OpenMP::in_parallel(OpenMP const& exec_space) noexcept {
+  return (
+      (exec_space.impl_internal_space_instance()->m_level < omp_get_level()) &&
+      (!Impl::t_openmp_instance ||
+       Impl::t_openmp_instance->m_level < omp_get_level()));
```
For those who like me struggle to see the diff
`Impl::OpenMPInternal::singleton().m_level -> exec_space.impl_internal_space_instance()->m_level`

Rombur replied:
If you hide whitespace it's pretty obvious what's changed.

dalg24 commented on core/src/OpenMP/Kokkos_OpenMP_Instance.hpp:
```diff
@@ -362,6 +365,48 @@ inline int OpenMP::impl_max_hardware_threads() noexcept {
+      std::is_arithmetic<T>::value,
+      "Kokkos Error: partitioning arguments must be integers or floats");
+  std::vector<OpenMP> instances(weights.size());
+  T total_weight = std::accumulate(weights.begin(), weights.end(), 0);
```
```suggestion
  T total_weight = std::accumulate(weights.begin(), weights.end(), T(0));
```
otherwise result is an `int`.

dalg24 commented on core/src/OpenMP/Kokkos_OpenMP_Instance.hpp:
```diff
@@ -362,6 +365,48 @@ inline int OpenMP::impl_max_hardware_threads() noexcept {
+  static_assert(
+      std::is_arithmetic<T>::value,
+      "Kokkos Error: partitioning arguments must be integers or floats");
+  std::vector<OpenMP> instances(weights.size());
```
I would prefer creating it empty and pushing (or emplacing) back in which case you can do a range-based for loop below

dalg24 commented on core/src/OpenMP/Kokkos_OpenMP_Instance.hpp:
```diff
@@ -362,6 +365,48 @@ inline int OpenMP::impl_max_hardware_threads() noexcept {
+
+  int resources_used = 0;
+  for (unsigned int i = 0; i < weights.size() - 1; ++i) {
+    int instance_pool_size = (weights[i] / total_weight) * main_pool_size;
```
What if it is `0`?  Are we delegating this to the instance constructor?
If that is not legal (haven't yet looked at the constructor) should we raise a warning/an error here to produce better diagnostic for the user?

Rombur replied:
We should do something for. I just realized that there is also a problem if you pass an empty weights vector

dalg24 replied:
The more I think about it, the more I am convinced that we should have a generic version that validates the input and then delegate to backend-specific implementations.  That way we can guarantee uniformity.  (Not asking you to do that as part of this PR.)

Rombur replied:
The other backends do not care about the inputs. They are thrown away. The only thing that matters is the number of weights

masterleinad replied:
Do you also want to check for `total_weight` not to be zero? Maybe, just check `weights[i]` instead of `instance_pool_size`?

masterleinad commented on core/src/OpenMP/Kokkos_OpenMP_Instance.hpp:
```diff
@@ -362,6 +363,62 @@ inline int OpenMP::impl_max_hardware_threads() noexcept {
+    if (instance_pool_size == 0) {
+      Kokkos::abort("Kokkos::abort: Instance has no resource allocated to it");
+    }
+    std::cout << "instance " << i << " " << instance_pool_size << " "
+              << weights[i] << " " << total_weight << " " << main_pool_size
+              << std::endl;
```
Do you intend to keep this output?

crtrott commented:
Worked a bit with a testcode to figure out whats going on and what should be going on:

```c++
#include <Kokkos_Core.hpp>
#include <cmath>

thread_local int id;
int id_counter;

double global_counter;

void busyloop(int N) {
  double tmp = 3.0;
  for(int i=0; i<N; i++)
    tmp += std::pow(tmp,0.9);
  global_counter += tmp;
}

int main(int argc, char* argv[]) {
  Kokkos::initialize(argc, argv);
  {
    int N = argc > 1 ? atoi(argv[1]) : 1000000;
    int R = argc > 2 ? atoi(argv[2]) : 10;

    #pragma omp parallel
    {
       id = 100 + Kokkos::atomic_fetch_inc(&id_counter);
    }

    {
      Kokkos::Timer timer;
      int num_threads = Kokkos::OpenMP().concurrency();
      #pragma omp parallel num_threads(4)
      {
        printf("omp num id nesting level: %i %i %i %i %i\n",int(omp_get_num_threads()),int(omp_get_thread_num()),id,int(omp_get_nested()),int(omp_get_level()));
        #pragma omp parallel num_threads(num_threads/4)
        {
          printf("nested_omp num id nesting level: %i %i %i %i %i\n",int(omp_get_num_threads()),int(omp_get_thread_num()),id,int(omp_get_nested()),int(omp_get_level()));
          busyloop(N);
        }
      }
      printf("time: %lf\n",timer.seconds());
    }

    auto instances = Kokkos::Experimental::partition_space(Kokkos::OpenMP(),0.25,0.25,0.25,0.25);
    {
      Kokkos::Timer timer;
      int num_threads = Kokkos::OpenMP().concurrency();
      #pragma omp parallel num_threads(4)
      {
        auto exec = instances[omp_get_thread_num()];
        printf("omp num id nesting level: %i %i %i %i %i %i\n",int(omp_get_num_threads()),int(omp_get_thread_num()),id,int(omp_get_nested()),int(omp_get_level()),exec.concurrency());
        Kokkos::parallel_for(Kokkos::RangePolicy<>(exec,0,num_threads/4),[=](int i)
        {
          printf("nested_omp num id nesting level: %i (%i %i %i) %i %i\n",int(omp_get_num_threads()),int(omp_get_thread_num()),i,id,int(omp_get_nested()),int(omp_get_level()));
          busyloop(N);
          //Kokkos::parallel_for(Kokkos::RangePolicy<>(exec,0,num_threads/4),[=](int i)
          //{});
        });
      }
      printf("time: %lf\n",timer.seconds());
    }

  }
  Kokkos::finalize();
}
```

Rombur commented:
@crtrott I have fixed `concurrency`. By default, the behavior is unchanged but you can pass an execution space if you have use `partition_space`. I have added a mutex and locks before accessing `m_pool` data. I don't think that the code can hang if you launch kernels from different threads. The kernels will be executed serially similar to what is done in the other backends.

Rombur commented:
@crtrott _ping_

masterleinad commented on core/src/OpenMP/Kokkos_OpenMP_Instance.hpp:
```diff
@@ -139,28 +147,24 @@ inline bool OpenMP::impl_is_initialized() noexcept {
+inline int OpenMP::impl_thread_pool_size(OpenMP const& exec_space) noexcept {
+  return OpenMP::in_parallel(exec_space)
              ? omp_get_num_threads()
              : (Impl::t_openmp_instance
```
In which cases do we actually need to use `t_openmp_instance`? Is that only necessary for `partition_master`?

Rombur replied:
Yes, it's only for `partition_master`

masterleinad commented on core/src/OpenMP/Kokkos_OpenMP_Instance.hpp:
```diff
@@ -362,6 +366,62 @@ inline int OpenMP::impl_max_hardware_threads() noexcept {
+    if (instance_pool_size == 0) {
+      Kokkos::abort("Kokkos::abort: Instance has no resource allocated to it");
+    }
+    std::cout << "instance " << i << " " << instance_pool_size << " "
+              << weights[i] << " " << total_weight << " " << main_pool_size
+              << std::endl;
```
```suggestion
```

masterleinad commented on core/src/OpenMP/Kokkos_OpenMP_Instance.hpp:
```diff
@@ -362,6 +366,62 @@ inline int OpenMP::impl_max_hardware_threads() noexcept {
+
+template <typename... Args>
+std::vector<OpenMP> partition_space(OpenMP const& main_instance, Args... args) {
+  // Unpack the arguments and create the weight vector. Note that if some of the
+  // types are not same, you will get a narrowing warning.
```
```suggestion
  // Unpack the arguments and create the weight vector. Note that if not all of the
  // types are the same, you will get a narrowing warning.
```

crtrott commented on core/src/Kokkos_OpenMP.hpp:
```diff
@@ -135,8 +137,7 @@ class OpenMP {
 
-  // use UniqueToken
-  static int concurrency();
+  static int concurrency(OpenMP const& = OpenMP());
```
Why can't this be a non-static member function as it kind of is supposed to be?

Rombur replied:
I did that for consistency. The function exists in several backends and it is always static.

crtrott replied:
OK we need an immediate follow up on this: this is ONLY static in the other guys because someone suggested that it is the same for every instance of currently existing types: it should NOT be static.

crtrott commented on core/src/Kokkos_OpenMP.hpp:
```diff
@@ -147,13 +148,12 @@ class OpenMP {
+  inline static int impl_thread_pool_rank() noexcept;
 
-  inline static int impl_thread_pool_size(int depth);
+  inline static int impl_thread_pool_size(int depth, OpenMP const& = OpenMP());
```
why not make this non-static?

Rombur replied:
I did that for consistency. The function exists in several backends and it is always static.

crtrott replied:
same as above

Rombur commented on core/src/Kokkos_OpenMP.hpp:
```diff
@@ -147,13 +148,12 @@ class OpenMP {
   static void impl_finalize();
 
-  inline static int impl_thread_pool_size() noexcept;
+  inline static int impl_thread_pool_size(OpenMP const& = OpenMP()) noexcept;
```
I did that for consistency. The function exists in several backends and it is always static.

crtrott replied:
same as above

dalg24 commented on core/unit_test/TestExecSpacePartitioning.hpp:
```diff
@@ -75,23 +76,61 @@ void check_distinctive(Kokkos::Experimental::SYCL exec1,
 #endif
+#ifdef KOKKOS_ENABLE_OPENMP
+void check_distinctive(Kokkos::OpenMP exec1, Kokkos::OpenMP exec2) {
+  ASSERT_NE(exec1.impl_internal_space_instance(),
+            exec2.impl_internal_space_instance());
```
```suggestion
  ASSERT_NE(exec1, exec2);
```

dalg24 commented on core/unit_test/TestExecSpacePartitioning.hpp:
```diff
@@ -75,23 +76,61 @@ void check_distinctive(Kokkos::Experimental::SYCL exec1,
 }
 #endif
+#ifdef KOKKOS_ENABLE_OPENMP
+void check_distinctive(Kokkos::OpenMP exec1, Kokkos::OpenMP exec2) {
```
Distinct or distintive?
Considering that you can directly compare them now, I would say drop the function.

dalg24 replied:
(OK now I see you didn't come up with the name.  Regardless I think we should get rid of that function.

Rombur replied:
I quickly tried to remove this function and it doesn't work out of the box. Let's do this in a different PR.

dalg24 commented on core/src/OpenMP/Kokkos_OpenMP_WorkGraphPolicy.hpp:
```diff
@@ -74,7 +74,11 @@ class ParallelFor<FunctorType, Kokkos::WorkGraphPolicy<Traits...>,
  public:
   inline void execute() {
-#pragma omp parallel num_threads(OpenMP::impl_thread_pool_size())
+    // Work around NVHPC 11.6 ICE
```
11.6 is a CUDA version.  Was it NVHPC 22.7?

Rombur replied:
No, I think it was 22.5

dalg24 commented on core/src/OpenMP/Kokkos_OpenMP_WorkGraphPolicy.hpp:
```diff
@@ -74,7 +74,11 @@ class ParallelFor<FunctorType, Kokkos::WorkGraphPolicy<Traits...>,
-#pragma omp parallel num_threads(OpenMP::impl_thread_pool_size())
+    // Work around NVHPC 11.6 ICE
+    int pool_size = OpenMP::impl_thread_pool_size();
+    // Work around HIP unused-variable warning
+    (void)pool_size;
```
Fun

dalg24 commented on core/src/OpenMP/Kokkos_OpenMP_Instance.hpp:
```diff
@@ -134,28 +142,24 @@ inline bool OpenMP::impl_is_initialized() noexcept {
-int OpenMP::impl_thread_pool_rank() noexcept {
-  // FIXME_OPENMP We are forced to use t_openmp_instance because the function is
-  // static
+inline int OpenMP::impl_thread_pool_rank() noexcept {
```
Did you mean to drop the `KOKKOS_INLINE_FUNCTION` annotation?  If so we probably should drop the `KOKKOS_IF_ON_{HOST,DEVICE}`

dalg24 commented on core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:
```diff
@@ -379,14 +381,32 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
     );
 
-    const int pool_size = OpenMP::impl_thread_pool_size();
+    if (OpenMP::in_parallel() &&
+        !(omp_get_nested() && (omp_get_level() == 1))) {
```
We check this condition a bunch of times.  Did you consider encapsulating it in a function?

PhilMiller commented on core/src/OpenMP/Kokkos_OpenMP_WorkGraphPolicy.hpp:
```diff
@@ -74,7 +74,11 @@ class ParallelFor<FunctorType, Kokkos::WorkGraphPolicy<Traits...>,
   inline void execute() {
-#pragma omp parallel num_threads(OpenMP::impl_thread_pool_size())
+    // Work around NVHPC 22.5 ICE
+    int pool_size = OpenMP::impl_thread_pool_size();
```
Why not `[[maybe_unused]]` instead of the workaround below?

Rombur replied:
Sure, I'll change it. The PR started 6 months ago so we couldn't use C++17

masterleinad commented on core/src/OpenMP/Kokkos_OpenMP_Instance.cpp:
```diff
@@ -194,10 +193,7 @@ void OpenMPInternal::resize_thread_data(size_t pool_reduce_bytes,
 
     memory_fence();
```
Also, can you please comment on the necessity to serialize the loop in the first place?

Rombur replied:
> Do you still need the memory fences here and below after serializing the loop?

I think we want to keep this memory_fence but we can get rid of the one in the loop.

> Also, can you please comment on the necessity to serialize the loop in the first place?

The problem is when you don't have nested parallelism, you only get `rank=0`. That's why I am doing the loop in serial to make sure that I really perform the loop.

Rombur commented on core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:
```diff
@@ -72,6 +72,11 @@
 namespace Kokkos {
 namespace Impl {
 
+inline bool in_serial() {
+  return (OpenMP::in_parallel() &&
+          !(omp_get_nested() && (omp_get_level() == 1)));
+}
```
The problem is not that it's nested. It's that you want to execute the code in parallel but the loop require nested parallelism which was not enable. I don't think that it should be called `is_nested` because that's not the issue itself. I can change it to `execute_in_serial` if you prefer.

masterleinad reviewed (approved):
Looks OK to me (assuming that there will be some cleanup for the static member functions later).

crtrott commented on core/src/OpenMP/Kokkos_OpenMP_Instance.hpp:
```diff
@@ -329,7 +333,7 @@ class UniqueToken<OpenMP, UniqueTokenScope::Global> {
   KOKKOS_INLINE_FUNCTION
   int acquire() const noexcept {
-    KOKKOS_IF_ON_HOST((return Kokkos::Impl::t_openmp_hardware_id;))
+    KOKKOS_IF_ON_HOST((return omp_get_thread_num();))
```
we need to discuss this in a follow up, because this needs to get a global thread id not the nesting level.

crtrott replied:
Can we add a fixme on this?

crtrott commented on core/src/OpenMP/Kokkos_OpenMP_Instance.hpp:
```diff
@@ -341,13 +345,13 @@ class UniqueToken<OpenMP, UniqueTokenScope::Global> {
 KOKKOS_INLINE_FUNCTION
 int OpenMP::impl_hardware_thread_id() noexcept {
-  KOKKOS_IF_ON_HOST((return Impl::t_openmp_hardware_id;))
+  KOKKOS_IF_ON_HOST((return omp_get_thread_num();))
```
this needs a bigger discussion

crtrott reviewed (approved):
Approve but needs a follow up.

dalg24 commented:
Retest this please

Rombur commented:
After the rebase on #5601, we are hitting https://github.com/kokkos/kokkos/issues/5651. There is also an error with HIP but it is spurious.

crtrott commented:
I see about 2-3% performance regression for LAMMPS with in.lj with size 4,2,2. This is on SKX with GCC 11.1 using 12 threads. 2.27s vs 2.32s.

crtrott commented:
I did some more experiments and implemented a different locking scheme here. Looks like mutex lock implies stronger memory fencing than we need: https://github.com/Rombur/kokkos/pull/4, and if we relax that a bit we recover the perf.

crtrott commented on core/src/OpenMP/Kokkos_OpenMP_Instance.cpp:
```diff
@@ -34,6 +34,23 @@
+                      desul::MemoryScopeDevice());
+}
+
+void release_lock() {}
```
what is this?

Rombur replied:
copy/paste error

vlkale commented on core/src/OpenMP/Kokkos_OpenMP_Instance.hpp:
```diff
@@ -70,6 +74,8 @@ class OpenMPInternal {
+
   static OpenMPInternal& singleton();
 
   void initialize(int thread_cound);
```
thread_count

cz4rs commented on core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:
```diff
@@ -199,19 +202,19 @@ class ParallelFor<FunctorType, Kokkos::MDRangePolicy<Traits...>,
   execute_parallel() const {
-#pragma omp parallel for schedule(static KOKKOS_OPENMP_OPTIONAL_CHUNK_SIZE) \
-    num_threads(OpenMP::impl_thread_pool_size())
+#pragma omp parallel for schedule(static, 1) \
```
Is that correct to just use `1` here?

Anyways, at least for the compilation to pass this change needs to be applied in line 193 as well.

cz4rs replied:
And if `1` is the correct value, it'd still be better to use named constant.

cz4rs commented on core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:
```diff
@@ -188,9 +191,9 @@ class ParallelFor<FunctorType, Kokkos::MDRangePolicy<Traits...>,
                                          Kokkos::Dynamic>::value>
   execute_parallel() const {
 #pragma omp parallel for schedule(dynamic KOKKOS_OPENMP_OPTIONAL_CHUNK_SIZE) \
```
Possibly:
```suggestion
#pragma omp parallel for schedule(dynamic, 1) \
```

crtrott replied:
Done

crtrott commented:
Here is the Serial fix corresponding to this for the Intel ICE: https://github.com/kokkos/kokkos/pull/5671
